### ISSUE TITLE
Implement lazy loading

### ISSUE THREAD

**Issue by babolivier** [matrix-org/dendrite#672]
Currently this isn't described in the specs. Let's make this wait till then.

**Comment by Kegsay** [comment]
More info please or I'll close this. Link to the relevant spec docs.

**Comment by Half-Shot** [comment]
Let me attempt to fill in some gaps.

So this came up as a [requirement to support Fluffychat](https://gitlab.com/ChristianPauly/fluffychat-flutter/-/issues/60), and I suspect going forwards a lot of client developers will come to rely on this feature as it's a performance improvement / bandwidth saver for the client. 

As such, rather than trying to convince client developers to support older spec versions, I'd rather push for Dendrite to support lazy loading.

**Comment by Half-Shot** [comment]
So links for the specs:

- Lazy loading room members spec - https://matrix.org/docs/spec/client_server/r0.6.0#lazy-loading-room-members
- /sync (RoomSummary section) https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-client-r0-sync
- /messages https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-client-r0-rooms-roomid-messages
- /context https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-client-r0-rooms-roomid-context-eventid

**Comment by Kegsay** [comment]
Sytests:
```
    × The only membership state included in an initial sync is for all the senders in the timeline
    × The only membership state included in an incremental sync is for senders in the timeline
    × The only membership state included in a gapped incremental sync is for senders in the timeline
    × Gapped incremental syncs include all state changes
    × Old leaves are present in gapped incremental syncs
    × Leaves are present in non-gapped incremental syncs
    × Old members are included in gappy incr LL sync if they start speaking
    × Members from the gap are included in gappy incr LL sync
    × We do send redundant membership state across incremental syncs if asked
    × Unnamed room comes with a name summary
    × Named room comes with just joined member count summary
    × Room summary only has 5 heroes
    × Room summary counts change when membership changes
    × GET /rooms/:room_id/messages lazy loads members correctly
```

**Comment by ShadowJonathan** [comment]
> `GET /rooms/:room_id/messages lazy loads members correctly`

I can confirm this one passes on dendrite now, maybe an update to this issue is warranted?

**Comment by neilalexander** [comment]
@S7evinK I think you did this for `/messages` — did you look at `/sync` also?

**Comment by S7evinK** [comment]
Currently it's only on `/context`, can have a look at `/messages` and `/sync`

**Comment by S7evinK** [comment]
Implemented in #2346, #2207 and #2218.
Will create a new PR where the cache from #2346 is used on `/messages` and `/context`.
